subsys/bluetooth/mesh/dfu_slot.c

```diff
@@ -339,16 +354,35 @@ static int slot_data_load(const char *key, size_t len_rd,
 
 	if (!strncmp(prop, PROP_HEADER, len)) {
 		if (read_cb(cb_arg, &slots[idx], HEADER_SIZE) > 0) {
-			sys_slist_append(&list, &slots[idx].n);
-		}
+			struct slot *s, *prev = NULL;
+
+			SYS_SLIST_FOR_EACH_CONTAINER(&list, s, n) {
+				if (s->idx > slots[idx].idx) {
+					break;
+				}
+
+				prev = s;
+			}
 
+			if (prev == NULL) {
+				sys_slist_prepend(&list, &slots[idx].n);
+			} else {
+				sys_slist_insert(&list, &prev->n, &slots[idx].n);
+			}
+
+			if (slots[idx].idx >= slot_index) {
+				slot_index = slots[idx].idx + 1;
+			}
+		}
 		return 0;
 	}
 
 	if (!strncmp(prop, PROP_FWID, len)) {
 		if (read_cb(cb_arg, &slots[idx].slot.fwid,
 			    sizeof(slots[idx].slot.fwid)) < 0) {
-			slot_invalidate(&slots[idx]);
+			slots[idx].idx = 0;
+			sys_slist_find_and_remove(&list, &slots[idx].n);
+			slot_erase(&slots[idx]);
```

Comment:
I'm not sure we can call `settings_delete()` from the `settings_load()`. If we reset `slot->idx` and remove the slot from `list`, can the slot be reused? If so, then we can skip the removal.